src/ObjLoading/Game/T6/AssetLoaders/AssetLoaderSoundBank.cpp

```diff
@@ -0,0 +1,584 @@
+
+    if (loadedEntryCount > 0)
+    {
+        sndBank->loadAssetBank.zone = memory->Dup(sndBankLocalization.at(0).c_str());
+        sndBank->loadAssetBank.language = memory->Dup(sndBankLocalization.at(1).c_str());
+        memset(sndBank->loadAssetBank.linkTimeChecksum, 0xCC, 16);
+
+        sndBank->loadedAssets.loadedCount = 0;
+        sndBank->loadedAssets.zone = memory->Dup(sndBankLocalization.at(0).c_str());
+        sndBank->loadedAssets.language = memory->Dup(sndBankLocalization.at(1).c_str());
+        sndBank->loadedAssets.entryCount = loadedEntryCount;
+        sndBank->loadedAssets.entries = static_cast<SndAssetBankEntry*>(memory->Alloc(sizeof(SndAssetBankEntry) * loadedEntryCount));
+        memset(sndBank->loadedAssets.entries, 0, sizeof(SndAssetBankEntry) * loadedEntryCount);
+
+        sndBank->runtimeAssetLoad = true;
+
+        const auto sablName = assetName + ".sabl";
+        sablStream = OpenSoundBankOutputFile(sablName);
+        if (sablStream)
+            sablWriter = SoundBankWriter::Create(sablName, *sablStream, searchPath);
+    }
+
+    if (streamedEntryCount > 0)
+    {
+        sndBank->streamAssetBank.zone = memory->Dup(sndBankLocalization.at(0).c_str());
+        sndBank->streamAssetBank.language = memory->Dup(sndBankLocalization.at(1).c_str());
+        memset(sndBank->streamAssetBank.linkTimeChecksum, 0xCC, 16);
+
+        const auto sabsName = assetName + ".sabs";
+        sabsStream = OpenSoundBankOutputFile(sabsName);
+        if (sabsStream)
+            sabsWriter = SoundBankWriter::Create(sabsName, *sabsStream, searchPath);
+    }
+
+    // add aliases to the correct sound bank writer
+    for (auto i = 0u; i < sndBank->aliasCount; i++)
+    {
+        auto* aliasList = &sndBank->alias[i];
+        for (auto j = 0; j < aliasList->count; j++)
+        {
+            auto* alias = &aliasList->head[j];
+
+            if (sabsWriter && alias->flags.loadType == T6::SA_STREAMED)
+                sabsWriter->AddSound(GetSoundFilePath(alias), alias->assetId, alias->flags.looping, true);
+            else if (sablWriter)
+                sablWriter->AddSound(GetSoundFilePath(alias), alias->assetId, alias->flags.looping);
+        }
+    }
+
+    // write the output linked sound bank
+    if (sablWriter)
+    {
+        auto size = static_cast<size_t>(sablWriter->Write());
+        sablStream->close();
+
+        if (size != UINT32_MAX)
+        {
+            sndBank->loadedAssets.dataSize = size;
+            sndBank->loadedAssets.data = static_cast<SndChar2048*>(memory->Alloc(size));
+            memset(sndBank->loadedAssets.data, 0, size);
+        }
+        else
+        {
+            std::cerr << "Loaded Sound Bank for " << assetName << " failed to generate. Please check your build files." << std::endl;
+            return false;
+        }
+    }
+
+    // write the output streamed sound bank
+    if (sabsWriter)
+    {
+        auto size = static_cast<size_t>(sabsWriter->Write());
```

Comment:
I agree the way that its organized currently doesn't make sense. I was originally planning to do something similar to what you suggested -- create a structure that while linking a zone would hold all of the sound data that needs to be written. Then after writing the zone out, the ZoneWriter could also write out all of the SoundBanks. My biggest concern for that was I didn't really like holding all the sound file data in memory if it wasn't necessary. It's probably not an issue since I think you already do something similar with the dumping and for images with ipaks so it would probably be okay. I can work on re-organizing now or after merged, whatever works best for you

Reply:
It will be changed somewhen else. I think this should just be merged for now since people can get value out of it, whether it is refactored or not.